### PR TITLE
Semgrep rules for adapters

### PR DESCRIPTION
**Pull request by onkarvhanumante**
- PR introduces semgrep rules for adapters code. These rules check whether,
   - response status codes are checked using utils
   - bid value in `typeBid` is assigned correctly
   - adapter.bidder interface is named correctly

- PR also adds unit tests for each semgrep rule. Note that semgrep looks for tests based on the rule filename and the languages specified in the rule. For example, we have rule for `no-content-status.yml` check. For this rule we have added `no-content-status.go`. This file includes patterns to be flagged by semgrep and patterns not to be flagged by semgrep.

- Note that this is initial step of integrating semgrep with adapter code reviews. As next step plan is to, 
   - Run rules for existing adapters and fix them.
   - Introduce GitHub action script that scans PR changes, run semgrep tests and add semgrep result as PR comment. Refer https://github.com/onkarvhanumante/prebid-server/pull/31. Therefore `message` field in each semgrep rule is written using markup language used for Github comment.

- Run semgrep unit tests 
```
~/prebid-repos/prebid-server (add-semgrep-rules*) » semgrep --test  /Users/oh0387/prebid-repos/prebid-server/.semgrep/adapter/                                                                                                                                                  
4/4: ✓ All tests passed
No tests for fixes found.
```

- Use following command to run semgrep tests against adapter code base
```
~/prebid-repos/prebid-server (add-semgrep-rules) » semgrep --gitlab-sast --config=.semgrep/adapter/ ./adapters/ | jq '[.vulnerabilities[] | {"file": .location.file, "start": .location.start_line, "end": .location.end_line, "message": (.message | gsub("\\n"; "\n"))}]' | jq -c


┌─────────────┐
│ Scan Status │
└─────────────┘
  Scanning 2492 files tracked by git with 4 Code rules:
  Scanning 191 files with 4 go rules.
  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00


┌──────────────┐
│ Scan Summary │
└──────────────┘
Some files were skipped or only partially analyzed.
  Scan was limited to files tracked by git.
  Scan skipped: 320 files matching .semgrepignore patterns
  For a full list of skipped files, run semgrep with the --verbose flag.

Ran 4 rules on 191 files: 477 findings.

A new version of Semgrep is available. See https://semgrep.dev/docs/upgrading
[{"file":"adapters/33across/33across.go","start":282,"end":282,"message":"Scope of `TtxAdapter` is limited to this adapter package. Therefore `TtxAdapter` can be renamed to `adapter`. Refer following example.\n```\n  type adapter struct {\n    endpoint string\n  }\n  func  Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server config.Server) (adapters.Bidder, error) {\n    return &adapter{endpoint: \"https://www.foo.com\"}, nil\n  }\n```\n"},{"file":"adapters/acuityads/acuityads.go","start":28,"end":28,"message":"Scope of `AcuityAdsAdapter` is limited to this adapter package. Therefore `AcuityAdsAdapter` can be renamed to `adapter`. Refer following example.\n```\n  type adapter struct {\n    endpoint string\n  }\n  func  Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server config.Server) (adapters.Bidder, error) {\n    return &adapter{endpoint: \"https://www.foo.com\"}, nil\n  }\n```\n"},{"file":"adapters/adgeneration/adgeneration.go","start":288,"end":288,"message":"Scope of `AdgenerationAdapter` is limited to this adapter package. Therefore `AdgenerationAdapter` can be renamed to `adapter`. Refer following example.\n```\n  type adapter struct {\n    endpoint string\n  }\n  func  Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server config.Server) (adapters.Bidder, error) {\n    return &adapter{endpoint: \"https://www.foo.com\"}, nil\n  }\n```\n"},{"file":"adapters/adhese/adhese.go","start":278,"end":278,"message":"Scope of `AdheseAdapter` is limited to this adapter pa
...
{"file":"adapters/videoheroes/videoheroes.go","start":137,"end":137,"message":"Found incorrect assignment made to `Bid` Recommended to either create copy of `bid` and then assign it as `Bid`  ```\n  for _, seatBid := range response.SeatBid {\n    for _, bid := range seatBids.Bid {\n      ...\n      seatBid := bid\n      responseBid := &adapters.TypedBid{\n        Bid: &seatBid,\n        ...\n      }\n      ...\n    }\n  }\n``` Alternatively can make use of `seatBids.Bid` loop index as shown below  ```\n  for _, seatBid := range response.SeatBid {\n    for i, _ := range seatBids.Bid {\n      ...\n      responseBid := &adapters.TypedBid{\n        Bid: &seatBids.Bid[i],\n        ...\n      }\n      ...\n      ...\n    }\n  }\n```\n"}]

~/prebid-repos/prebid-server (add-semgrep-rules) »

```

### CODE REVIEW

File: .semgrep/adapter/type-bid-assignment.yml [hunk 1]
```diff
@@ -0,0 +1,61 @@
+rules:
+  - id: type-bid-assignment-check
+    languages:
+      - go
+    message: >
+      Found incorrect assignment made to `Bid` Recommended to either create copy of `bid` and then assign it as `Bid`  
+
+      ```
+        for _, seatBid := range response.SeatBid {
+          for _, bid := range seatBids.Bid {
```
Comment (Sonali-More-Xandr) [inline .semgrep/adapter/type-bid-assignment.yml:10]: nitpick:- typo here it should be `seatBid.Bid`

File: .semgrep/adapter/type-bid-assignment.yml [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            ...
+            seatBid := bid
```
Comment (Sonali-More-Xandr) [inline .semgrep/adapter/type-bid-assignment.yml:12]: Please rename this copy variable to `bidCopy` maybe as we already have a range variable named `seatBid`

File: .semgrep/adapter/type-bid-assignment.yml (hunk 1 above)
Comment (onkarvhanumante) [reply .semgrep/adapter/type-bid-assignment.yml]: [e49d669](https://github.com/prebid/prebid-server/pull/2833/commits/e49d669a214a88ded6094d03bf4ea1760b4f63c5) fixes typo

**Review by Sonali-More-Xandr — APPROVED** [review]
LGTM!

File: .semgrep/adapter/type-bid-assignment.yml [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+          - pattern: >
+              for _, $BID := range ... {
+                ...
+                ... = adapters.TypedBid{
+                  $KEY: &$BID,
+                  ...
+                }
+                ...
+              } 
+          - pattern: >
```
Comment (mansinahar) [inline .semgrep/adapter/type-bid-assignment.yml:42]: Won't we need another case here as 
```
          - pattern: >
              for _, $BID := range ... {
                ...
                ... := append(...,  &adapters.TypedBid{
                  $KEY: &$BID,
                  ...
                })
                ...
              }
```

File: .semgrep/adapter/type-bid-assignment.yml [hunk 4]
```diff
@@ -0,0 +1,61 @@
+rules:
+  - id: type-bid-assignment-check
+    languages:
+      - go
+    message: >
+      Found incorrect assignment made to `Bid` Consider using `seatBids.Bid` loop index as shown below
```
[the remark this answers is not in the document]
Comment (onkarvhanumante) [reply .semgrep/adapter/type-bid-assignment.yml]: https://github.com/prebid/prebid-server/pull/2833/commits/dc62fffd8a7eff6d77172cad9cd78ea5361a9615 adds the reasoning behind not assigning loop variable pointer value as bid

File: .semgrep/adapter/builder-struct-name.yml [hunk 5]
```diff
@@ -0,0 +1,55 @@
+rules:
+  - id: builder-struct-name-check
+    languages:
+      - go
+    message: >
+      Scope of `$BUILDER` is limited to this adapter package. Therefore `$BUILDER` can be renamed to `adapter`. Refer following example.
```
[the remark this answers is not in the document]
Comment (onkarvhanumante) [reply .semgrep/adapter/builder-struct-name.yml]: https://github.com/prebid/prebid-server/pull/2833/commits/f25485c461aa1a9f31b89647df0a40e11fc2b983

[the remark this answers is not in the document]
Comment (onkarvhanumante) [reply .semgrep/adapter/builder-struct-name.yml]: https://github.com/prebid/prebid-server/pull/2833/commits/03ee1380b1fbaa96c0321ce60d900c81cf5f06c6 updates message

File: .semgrep/adapter/type-bid-assignment.yml (hunk 3 above)
Comment (onkarvhanumante) [reply .semgrep/adapter/type-bid-assignment.yml:45]: we won't have `:=  append` in loop

Comment (mansinahar) [reply .semgrep/adapter/type-bid-assignment.yml:45]: Ah ok! Yeah, you're right

File: .semgrep/adapter/type-bid-assignment.yml [hunk 6]
```diff
@@ -2,8 +2,11 @@ rules:
   - id: type-bid-assignment-check
     languages:
       - go
-    message: >
-      Found incorrect assignment made to `Bid` Consider using `seatBids.Bid` loop index as shown below  
+    message: >      
+      Found incorrect assignment made to $KEY. $BID variable receives a new value in each iteration of range loop. Assigning the address of $BID `(&$BID)` to $KEY will result in a pointer that always points to the same memory address with the value of the last iteration. 
+      This can lead to unexpected behavior or incorrect results. Refer https://go.dev/play/p/9ZS1f-5h4qS
+
+      Consider using `seatBids.Bid` loop index as shown below
```
Comment (mansinahar) [inline .semgrep/adapter/type-bid-assignment.yml:9]: **Nitpick:** Consider rewriting the last sentence as:
`Consider using an index variable in the seatBids.Bid loop as shown below`

File: .semgrep/adapter/builder-struct-name.yml [hunk 7]
```diff
@@ -3,9 +3,22 @@ rules:
     languages:
       - go
     message: >
-      Scope of `$BUILDER` is limited to this adapter package. Therefore `$BUILDER` can be renamed to `adapter`. Refer to the following example.
+      Using the adapter name for the struct makes it redundant with the package name.
```
Comment (mansinahar) [inline .semgrep/adapter/builder-struct-name.yml:6]: I would recommend making this comment a little simpler such as:
```
Using $BUILDER for this struct name makes it redundant with the package name. Consider renaming this struct to "adapter". See example below:
```

**Review by mansinahar — COMMENTED** [review]
@onkarvhanumante thanks for making the changes. I just have two minor comments about the messages.

File: .semgrep/adapter/builder-struct-name.yml (hunk 7 above)
Comment (onkarvhanumante) [reply .semgrep/adapter/builder-struct-name.yml]: https://github.com/prebid/prebid-server/pull/2833/commits/bd7aaffb9104aeda884a420f785bbdcef74786c3

thanks for the updated message

File: .semgrep/adapter/type-bid-assignment.yml (hunk 6 above)
Comment (onkarvhanumante) [reply .semgrep/adapter/type-bid-assignment.yml]: https://github.com/prebid/prebid-server/pull/2833/commits/bd7aaffb9104aeda884a420f785bbdcef74786c3

**Review by mansinahar — APPROVED** [review]

**Comment by Sonali-More-Xandr** [comment]
I see that the HTTP response checks were removed from this PR. I agree that the use of utils is optional and should not block the PR from merging. It is upto the author if they want to update the adapter to use the utils or not but it is better to offer them an option. I think we should keep a semgrep rule that will add a comment as optional and let the author make decision.
thoughts?

**Comment by mansinahar** [comment]
@Sonali-More-Xandr I don't have a very strong opinion on this one so if you both think it will be valuable to have that as a warning, I won't fight it :)

**Comment by onkarvhanumante** [comment]
> I see that the HTTP response checks were removed from this PR. I agree that the use of utils is optional and should not block the PR from merging. It is upto the author if they want to update the adapter to use the utils or not but it is better to offer them an option. I think we should keep a semgrep rule that will add a comment as optional and let the author make decision. thoughts?



> @Sonali-More-Xandr I don't have a very strong opinion on this one so if you both think it will be valuable to have that as a warning, I won't fight it :)


For now lets have 2 semgrep rules. We can add other in new PR

**Review by Sonali-More-Xandr — APPROVED** [review]

**Comment by shahinrahbariasl** [comment]
Hey, our gofmt job in our build pipeline started failing due to missing package keyword in some of these files. Is that intended? Shouldn't all go files have a package declaration?

**Pull request merged**